### PR TITLE
[core] Support stream read with delay duration

### PR DESCRIPTION
**Pull request by xuzifu666**
### Purpose
Support stream read with delay duration.
real user scenario is algorithm partital update, user use lake table join hbase/redis dim table, but dim table update is delay, so need delay read for stream lake table. this pr is aim to support it with add an option to control the delay time.

Linked issue: close #xxx

### CODE REVIEW

File: docs/layouts/shortcodes/generated/core_configuration.html [hunk 1]
```diff
@@ -647,6 +647,12 @@
             <td>Long</td>
             <td>Optional snapshot id used in case of "from-snapshot" or "from-snapshot-full" scan mode</td>
         </tr>
+        <tr>
+            <td><h5>scan.stream.delay.read.duration</h5></td>
```
Comment (JingsongLi) [inline docs/layouts/shortcodes/generated/core_configuration.html:651]: `streaming-read-snapshot-delay`?

File: docs/layouts/shortcodes/generated/core_configuration.html [hunk 2]
```diff
@@ -647,6 +647,12 @@
...
+            <td style="word-wrap: break-word;">(none)</td>
+            <td>Duration</td>
+            <td>The delay duration of stream read when scan snapshots</td>
```
Comment (JingsongLi) [inline docs/layouts/shortcodes/generated/core_configuration.html:654]: The delay duration of stream read when scan incremental snapshots.

File: paimon-core/src/main/java/org/apache/paimon/table/source/DataTableStreamScan.java [hunk 3]
```diff
@@ -198,6 +202,19 @@ private Plan nextPlan() {
         }
     }
 
+    private boolean checkDelaySnapshot(long snapshotId) {
+        if (options.scanDelayDuration() == null) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/table/source/DataTableStreamScan.java:206]: extract a class field: `Long scanDelayMillis`.

File: docs/layouts/shortcodes/generated/core_configuration.html (hunk 1 above)
Comment (xuzifu666) [reply docs/layouts/shortcodes/generated/core_configuration.html]: Done,  thanks~ @JingsongLi

File: docs/layouts/shortcodes/generated/core_configuration.html (hunk 2 above)
Comment (xuzifu666) [reply docs/layouts/shortcodes/generated/core_configuration.html]: Done

File: paimon-core/src/main/java/org/apache/paimon/table/source/DataTableStreamScan.java (hunk 3 above)
Comment (xuzifu666) [reply paimon-core/src/main/java/org/apache/paimon/table/source/DataTableStreamScan.java]: Done, had extract to a class filed.

File: paimon-core/src/main/java/org/apache/paimon/table/source/DataTableStreamScan.java [hunk 4]
```diff
@@ -177,6 +182,10 @@ private Plan nextPlan() {
                 throw new EndOfScanException();
             }
 
+            if (checkDelaySnapshot(nextSnapshotId)) {
+                continue;
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/table/source/DataTableStreamScan.java:186]: There will be a busy loop?

Can we just return `SnapshotNotExistPlan.INSTANCE` here?

File: paimon-core/src/main/java/org/apache/paimon/table/source/DataTableStreamScan.java [hunk 5]
```diff
@@ -198,6 +207,19 @@ private Plan nextPlan() {
         }
     }
 
+    private boolean checkDelaySnapshot(long snapshotId) {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/table/source/DataTableStreamScan.java:210]: `shouldDelaySnapshot`?

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java [hunk 6]
```diff
@@ -1846,6 +1853,10 @@ public Integer scanManifestParallelism() {
         return options.get(SCAN_MANIFEST_PARALLELISM);
     }
 
+    public Duration scanDelayDuration() {
```
[the remark this answers is not in the document]
Comment (xuzifu666) [reply paimon-common/src/main/java/org/apache/paimon/CoreOptions.java]: done

File: paimon-core/src/main/java/org/apache/paimon/table/source/DataTableStreamScan.java (hunk 5 above)
Comment (xuzifu666) [reply paimon-core/src/main/java/org/apache/paimon/table/source/DataTableStreamScan.java]: done.

File: paimon-common/src/main/java/org/apache/paimon/CoreOptions.java [hunk 7]
```diff
@@ -658,6 +658,13 @@ public class CoreOptions implements Serializable {
                                     + "Note: Scale-up this parameter will increase memory usage while scanning manifest files. "
                                     + "We can consider downsize it when we encounter an out of memory exception while scanning");
 
+    public static final ConfigOption<Duration> SCAN_DELAY_READ_DURATION =
```
[the remark this answers is not in the document]
Comment (xuzifu666) [reply paimon-common/src/main/java/org/apache/paimon/CoreOptions.java]: done

**Comment by JingsongLi** [comment]
+1

**Pull request merged**